torch/csrc/jit/codegen/cuda/kernel_ir.h

```diff
@@ -23,26 +24,276 @@ namespace cuda {
+//! (Passkey::kernel)
 //!
-//! Granular "friendship" token, used to implement the "passkey" idiom:
+//! It is a "granular friendship" token, used to implement the "passkey" idiom:
 //! https://www.spiria.com/en/blog/desktop-software/passkey-idiom-and-better-friendship-c
 //! https://arne-mertz.de/2016/10/passkey-idiom
 //!
 class Passkey {
   friend class IrBuilder;
-  Passkey() {}
+
+ public:
+  Kernel* const kernel = nullptr;
+
+ private:
+  explicit Passkey(Kernel* kernel) : kernel(kernel) {}
+};
+
+//! Kernel IR visitor interface
+class TORCH_CUDA_API IrVisitor : public PolymorphicBase {
+ public:
+  // TODO(kir): use Node* instead of void*
+  virtual void unhandled(const void* node) {}
+
+  // Values
+  virtual void visit(const NamedScalar* named_scalar) {
+    unhandled(named_scalar);
+  }
+  virtual void visit(const Bool* value) {
+    unhandled(value);
+  }
+  virtual void visit(const Float* value) {
+    unhandled(value);
+  }
+  virtual void visit(const Half* value) {
+    unhandled(value);
+  }
+  virtual void visit(const Int* value) {
+    unhandled(value);
+  }
+  virtual void visit(const IterDomain* iter_domain) {
+    unhandled(iter_domain);
+  }
+  virtual void visit(const TensorDomain* tensor_domain) {
+    unhandled(tensor_domain);
+  }
+  virtual void visit(const TensorView* tensor_view) {
+    unhandled(tensor_view);
+  }
+  virtual void visit(const TensorIndex* tensor_index) {
+    unhandled(tensor_index);
+  }
+
+  // Expressions
+  virtual void visit(const UnaryOp* node) {
+    unhandled(node);
+  }
+  virtual void visit(const BinaryOp* node) {
+    unhandled(node);
+  }
+  virtual void visit(const TernaryOp* node) {
+    unhandled(node);
+  }
+  virtual void visit(const ReductionOp* node) {
+    unhandled(node);
+  }
+  virtual void visit(const BroadcastOp* node) {
+    unhandled(node);
+  }
+
+  // Statements
+  virtual void visit(const Allocate* node) {
+    unhandled(node);
+  }
+  virtual void visit(const Sync* node) {
+    unhandled(node);
+  }
+  virtual void visit(const ForLoop* node) {
+    unhandled(node);
+  }
+  virtual void visit(const IfThenElse* node) {
+    unhandled(node);
+  }
+  virtual void visit(const GridReduction* node) {
+    unhandled(node);
+  }
+};
+
+//! Base class for Kernel IR nodes
+class TORCH_CUDA_API Node : public NonCopyable, public PolymorphicBase {
+ public:
+  explicit Node(Passkey) {}
+
+  //! IR Visitor double-dispatch interface
+  //! (https://en.wikipedia.org/wiki/Visitor_pattern)
+  virtual void accept(IrVisitor* visitor) const = 0;
+};
+
+//! Generic value (scalar or tensor)
+class TORCH_CUDA_API Val : public Node {
+ public:
+  Val(Passkey passkey, DataType dtype);
+
+  // TODO(kir): consider renaming
+  StmtNameType name() const {
+    return name_;
+  }
+
+  void setName(StmtNameType name) {
+    name_ = name;
+  }
+
+  ValueId id() const {
+    return id_;
+  }
+
+  DataType dtype() const {
+    return dtype_;
+  }
+
+  Expr* definition() const {
```

Comment:
Can we use the same terminology between the Fusion and Kernel IRs? Either sticking to `definition` or `origin`.

Reply:
Yes, the intention is to switch to definition for fusion too, but in a separate PR - "use/def" is common compiler terminology.